Fix torch profiler bugs for bench_offline_throughput.py

PanJason opened the pull request:
<!-- Thank you for your contribution! We appreciate it. The following guidelines will help improve your pull request and facilitate feedback. If anything is unclear, don't hesitate to submit your pull request and ask the maintainers for assistance. -->

## Motivation

This set of patches fixes some bugs using torch profiler in bench_offline_throughput.py

## Modifications

1. The first commit temporarily disables stack usage for pytorch profiler, reported here https://github.com/vllm-project/vllm/issues/18240 and https://github.com/sgl-project/sglang/issues/6236 . We can change it back when we upgrade the torch version
2. The second commit adds the `start_profile` and `stop_profile` methods for `Runtime` so both `Engine` and `Runtime` can use the same method in the script
3. The third commit changes the method used to create a new process from fork to spawn to fix the problem of calling `get_mm_processor` in `tokenizer_manager, so that the CUDA context can initialize correctly.
4.  The fourth commit fixes a deadlock of `monitor_trace_file` as `stop_profile` can already generate the file before `monitor_trace_file` is called.
5.  The fifth commit fixes the handling of different return value types of `get_server_info`. Runtime's method is async while `Engine`'s method is sync.

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

PanJason committed:
Disable stack due to a pytorch bug

PanJason committed:
Add start_ and end_profiler for Runtime.

So bench_offline_profiler.py can use torch profiler for both backends

PanJason committed:
Change the launch_server method from fork to spawn

To fix the following error when get_mm_processors()
```bash
Ignore import error when loading \
sglang.srt.managers.multimodal_processors.qwen_vl: Cannot re-initialize\
CUDA in forked subprocess. To use CUDA with multiprocessing, you must\
use the 'spawn' start method
```

PanJason committed:
Fix a deadlock in monitor_trace_file

PanJason committed:
Runtime's get_server_info method is async

xiezhq-hermann commented on python/sglang/srt/managers/tokenizer_manager.py:
```diff
@@ -800,7 +800,7 @@ async def start_profile(
             num_steps=num_steps,
             activities=activities,
-            with_stack=with_stack,
+            with_stack=False,
```
does this change still applicable with the current version? also if this is not an urgent fix maybe we should not make this change.

xiezhq-hermann commented:
@hnyls2002 can you help take a look at the changes on the profiler part?

hnyls2002 commented:
https://github.com/pytorch/pytorch/issues/113564 I think this problem may be related to a pytorch internal issue. I suggest making `with_stack` set to true by default but can be set by an env var.

PanJason committed:
Change with_stack option to use env var

Following the suggestion to use env var
SGLANG_PROFILE_WITH_STACK
with the original `with_stack` option to decide
whether setting stack for pytorch profiler
or not.
The expected behaviour:
If both `with_stack` and SGLANG_PROFLIE_WITH_STACK
are not set, the `with_stack` option is kept as None,
otherwise False if any of these is False.

PanJason committed:
Add bug info to profile doc

hnyls2002 commented on python/sglang/srt/managers/tokenizer_manager.py:
```diff
@@ -803,6 +803,17 @@ async def start_profile(
+            if env_with_stack_str is None
+            else env_with_stack_str.lower() in ["true", "1"]
+        )
+        with_stack = (
```
Set `with_stack=True` by default; I think that, currently, in most cases, profiling with a stack would not crash.

PanJason committed:
Change the default and default of env to True

hnyls2002 committed:
Merge branch 'main' into torch_profiler_fix
